### PR TITLE
stricten union types + add ci and basic tests

### PR DESCRIPTION
**Pull request by n1ru4l**
This addresses the last paragraph mentioned in https://github.com/sikanhe/gqtx/issues/9

I updated TypeScript to 3.9.x as 3.9 introduced [`@ts-expect-error` comments](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-9.html#-ts-expect-error-comments).

This change assumes that the `Src` of a `Union` is always the typescript union type of the single object types. It would solve my use-case. 

Previously only the resolved type was returned from that function. This could result in the object passed into the resolved object types field resolvers being incorrect. By changing the API to return a tuple with both the object type and the object the compiler can ensure that the object is actually compatible with the Src type of the object type. It does not really have an impact on runtime. 

See the tests: https://github.com/sikanhe/gqtx/pull/12/files#diff-8d994fbf3724bd6655cf9e633fceced3

I also added a GitHub action that compiles the project + checks the test typings.

@sikanhe Would love to hear your feedback (+ @FedericoBiccheddu, as you are also contributing back to this project!)

### CODE REVIEW

File: test/type-test.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
+import { createTypesFactory } from "../src/index";
```
Comment (n1ru4l) [inline test/type-test.ts:1]:
```suggestion
import { createTypesFactory } from "../dist/index";
```
Changing this line breaks the tests.

When changing this line the type of the first argument passed to the `resolveType` changes from `A | B` to `A`. I currently do not have any idea why.

**Comment by n1ru4l** [comment]
After thinking some more about this I wonder whether this actually makes sense. This won't help when people return objects that are not derived from the argument passed to `resolveType`. Maybe there is a better solution to this problem?

Maybe it makes sense to actually pass the returned object to the actual resolver (if possible).

File: tsconfig.json [hunk 2]
```diff
@@ -1,24 +1,18 @@
 {
   "compilerOptions": {
-    "incremental": true,                   /* Enable incremental compilation */
```
Comment (sikanhe) [inline tsconfig.json:3]: Downside of removing incremental?

**Comment by FedericoBiccheddu** [comment]
Hi @n1ru4l, thank you for considering me! 

I would not bump the TS's version because would make the package more difficult to adopt; not everyone uses the latest version, but I see the point here. Maybe can I help you to introduce some tests and leave the current TS version if it solves the problem?

File: tsconfig.json (hunk 2 above)
Comment (n1ru4l) [reply tsconfig.json:3]: I had issues compiling the project. Some files that I definitely changed were not rebuilt. I can change it back but for development purposes, I think it is a bit cumbersome.

**Comment by n1ru4l** [comment]
Gonna close this for now, until there is more input in https://github.com/sikanhe/gqtx/issues/21

**Pull request closed without merging**